sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ImplementISerializableCorrectly.cs

```diff
@@ -0,0 +1,246 @@
+                yield break;
+            }
+
+            var constructorSyntax = GetDeclarations<ConstructorDeclarationSyntax>(serializationConstructor).First();
+
+            if (typeSymbol.IsSealed && serializationConstructor.DeclaredAccessibility != Accessibility.Private ||
+                !typeSymbol.IsSealed && serializationConstructor.DeclaredAccessibility != Accessibility.Protected)
+            {
+                yield return new SecondaryLocation(constructorSyntax.Identifier.GetLocation(),
+                    $"Make this constructor '{accessibility}'.");
+            }
+
+            if (ImplementsISerializable(typeSymbol.BaseType) &&
+                !IsCallingBaseConstructor(serializationConstructor))
+            {
+                yield return new SecondaryLocation(constructorSyntax.Identifier.GetLocation(),
+                    $"Call constructor 'base(SerializationInfo, StreamingContext)'.");
+            }
+        }
+
+        private static bool IsCallingBase(IMethodSymbol methodSymbol, SemanticModel semanticModel)
+        {
+            var methodDeclaration = (MethodDeclarationSyntax)methodSymbol
+                .DeclaringSyntaxReferences
+                .FirstOrDefault()
+                ?.GetSyntax();
+            if (methodDeclaration == null)
+            {
+                return false;
+            }
+
+            return methodDeclaration.DescendantNodes()
+                .OfType<InvocationExpressionSyntax>()
+                .Select(i => semanticModel.GetSymbolInfo(i).Symbol)
+                .OfType<IMethodSymbol>()
+                .Any(m => m.ContainingType.Equals(methodSymbol.ContainingType.BaseType) &&
+                          KnownMethods.IsGetObjectData(m));
+        }
+
+        private static bool IsCallingBaseConstructor(IMethodSymbol constructorSymbol)
+        {
+            var constructorDeclaration = (ConstructorDeclarationSyntax)constructorSymbol
+                .DeclaringSyntaxReferences
+                .FirstOrDefault()
+                ?.GetSyntax();
+            if (constructorDeclaration == null)
+            {
+                return false;
+            }
+
+            var baseKeyword = constructorDeclaration?.Initializer?.ThisOrBaseKeyword;
+            return baseKeyword.HasValue && baseKeyword.Value.IsKind(SyntaxKind.BaseKeyword);
+        }
+
+        private static bool ImplementsISerializable(ITypeSymbol typeSymbol)
+        {
+            return typeSymbol != null &&
+                typeSymbol.IsPublicApi() &&
+                typeSymbol.AllInterfaces.Any(IsOrImplementsISerializable);
+        }
+
+        private static bool HasSerializableAttribute(ITypeSymbol typeSymbol)
+        {
+            return typeSymbol.GetAttributes()
+                .Any(a => a.AttributeClass.Is(KnownType.System_SerializableAttribute));
+        }
+
+        private static bool IsOrImplementsISerializable(ITypeSymbol typeSymbol)
+        {
+            return typeSymbol != null &&

```

Comment:
Not sure if you are obliged to test to null first here because:
- Is and Implements already do it
- the calling method do it too